File: python/src/aoc/year2023/day19.py
import re

from aoc.util import load_input, load_example


class Part:

    PART_PATTERN = re.compile(r"\{x=(\d+),m=(\d+),a=(\d+),s=(\d+)}")

    def __init__(self, line):
        x, m, a, s = map(int, self.PART_PATTERN.match(line).groups())
        self.registers = {"x": x, "m": m, "a": a, "s": s}

    def rating(self):
        return sum(self.registers.values())


class LessThanRule:
    def __init__(self, variable, value, next_workflow):
        self.variable = variable
        self.value = int(value)
        self.next_workflow = next_workflow

    def match(self, part):
        return part.registers[self.variable] < self.value

    def anti(self):
        return GreaterThanRule(self.variable, self.value - 1, "anti")


class GreaterThanRule:
    def __init__(self, variable, value, next_workflow):
        self.category = variable
        self.value = int(value)
        self.next_workflow = next_workflow

    def match(self, part):
        return part.registers[self.category] > self.value

    def anti(self):
        return LessThanRule(self.category, self.value + 1, "anti")


class DirectRule:
    def __init__(self, next_workflow):
        self.next_workflow = next_workflow

    def match(self, _):
        return True


class Workflow:

    LESS_THAN_PATTERN = re.compile(r"([xmas])<(\d+):([ARa-z]+)")
    GREATER_THAN_PATTERN = re.compile(r"([xmas])>(\d+):([ARa-z]+)")

    def __init__(self, rules):
        self.rules = []
        for rule in rules.split(","):
            if self.LESS_THAN_PATTERN.match(rule):
                self.rules.append(LessThanRule(*self.LESS_THAN_PATTERN.match(rule).groups()))
            elif self.GREATER_THAN_PATTERN.match(rule):
                self.rules.append(GreaterThanRule(*self.GREATER_THAN_PATTERN.match(rule).groups()))
            elif rule == "A" or rule == "R":
                self.rules.append(DirectRule(rule))
            elif re.match(r"[a-z]+", rule):
                self.rules.append(DirectRule(rule))
            else:
                raise

    def process(self, part):
        for rule in self.rules:
            if rule.match(part):
                return rule.next_workflow
        raise


class Workflows:

    WORKFLOW_PATTERN = re.compile(r"([a-z]+)\{(.*)}")

    def __init__(self, lines):
        it = iter(lines)

        self.workflows = {}
        for line in it:
            if not line:
                break
            g = self.WORKFLOW_PATTERN.match(line)
            name, rules = g.groups()
            self.workflows[name] = Workflow(rules)

        self.parts = [Part(line) for line in it]

    def _handle_part(self, part):
        current_workflow = self.workflows["in"]
        while True:
            result = current_workflow.process(part)
            if result in ("A", "R"):
                return result
            current_workflow = self.workflows[result]

    def sum_accepted_parts(self):
        return sum(part.rating() for part in self.parts if self._handle_part(part) == "A")

    @staticmethod
    def _upper_bound(constraints, category):
        values = [rule.value for rule in constraints if isinstance(rule, LessThanRule) and rule.variable == category]
        return min(values) if values else 4001

    @staticmethod
    def _lower_bound(constraints, category):
        values = [rule.value for rule in constraints if isinstance(rule, GreaterThanRule) and rule.category == category]
        return max(values) if values else 0

    def _collect_result(self, collected_rules):
        total = 1
        for v in "xmas":
            lb, ub = self._lower_bound(collected_rules, v), self._upper_bound(collected_rules, v)
            assert lb < ub
            p = ub - lb - 1
            total *= p
        return total

    def dfs(self, current_node="in", i=0, constraints=None):
        if constraints is None:
            constraints = []
        if current_node == "A":
            return self._collect_result(constraints)
        if current_node == "R":
            return 0
        current_workflow = self.workflows[current_node]
        if i >= len(current_workflow.rules):
            raise
        current_rule = current_workflow.rules[i]
        # match
        total = self.dfs(current_rule.next_workflow, 0, constraints + [current_rule])
        # no match
        if i + 1 < len(current_workflow.rules):
            total += self.dfs(current_node, i + 1, constraints + [current_rule.anti()])
        return total


def part1(lines):
    """
    >>> part1(load_example(__file__, "19"))
    19114
    """
    workflows = Workflows(lines)
    return workflows.sum_accepted_parts()


def part2(lines):
    """
    >>> part2(load_example(__file__, "19"))
    167409079868000
    """
    workflows = Workflows(lines)
    return workflows.dfs()


if __name__ == "__main__":
    data = load_input(__file__, 2023, "19")
    print(part1(data))
    print(part2(data))

File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2023/Day19.kt
package io.github.ocirne.aoc.year2023

import io.github.ocirne.aoc.AocChallenge

class Day19(val lines: List<String>) : AocChallenge(2023, 19) {

    companion object {
        private val IDENTIFIER_PATTERN = """[a-z]+""".toRegex()
        private val END_STATE_PATTERN = """[AR]""".toRegex()
        private val PART_PATTERN = """\{x=(\d+),m=(\d+),a=(\d+),s=(\d+)}""".toRegex()
        private val WORKFLOW_PATTERN = """([a-z]+)\{(.*)}""".toRegex()
        private val LESS_THAN_PATTERN = """([xmas])<(\d+):([ARa-z]+)""".toRegex()
        private val GREATER_THAN_PATTERN = """([xmas])>(\d+):([ARa-z]+)""".toRegex()
    }

    data class Part(val line: String) {

        val registers =
            "xmas".toCharArray()
                .zip(PART_PATTERN.find(line)!!.groupValues.drop(1).map { it.toInt() })
                .toMap()

        fun rating(): Int = registers.values.sum()
    }

    abstract class Rule(val nextWorkflow: String) {
        abstract fun match(part: Part): Boolean

        abstract fun anti(): Rule
    }

    class LessThanRule(val variable: Char, val value: Int, nextWorkflow: String) : Rule(nextWorkflow) {

        override fun match(part: Part): Boolean = part.registers[variable]!! < value

        override fun anti(): Rule = GreaterThanRule(variable, value - 1, "anti")
    }

    class GreaterThanRule(val variable: Char, val value: Int, nextWorkflow: String) :
        Rule(nextWorkflow) {

        override fun match(part: Part): Boolean = part.registers[variable]!! > value

        override fun anti(): Rule = LessThanRule(variable, value + 1, "anti")
    }

    class DirectRule(nextWorkflow: String) : Rule(nextWorkflow) {

        override fun match(part: Part): Boolean = true

        override fun anti(): Rule = this
    }

    class Workflow(rulesString: String) {

        val rules = rulesString.split(",").map { rule ->
            when {
                LESS_THAN_PATTERN.matches(rule) -> {
                    val (variable, value, nextWorkflow) = LESS_THAN_PATTERN.find(rule)!!.destructured
                    LessThanRule(variable.first(), value.toInt(), nextWorkflow)
                }

                GREATER_THAN_PATTERN.matches(rule) -> {
                    val (variable, value, nextWorkflow) = GREATER_THAN_PATTERN.find(rule)!!.destructured
                    GreaterThanRule(variable.first(), value.toInt(), nextWorkflow)
                }

                END_STATE_PATTERN.matches(rule) ->
                    DirectRule(rule)

                IDENTIFIER_PATTERN.matches(rule) ->
                    DirectRule(rule)

                else ->
                    throw IllegalStateException(rule)
            }
        }

        fun process(part: Part): String = rules.first { it.match(part) }.nextWorkflow
    }

    class Workflows(val lines: List<String>) {

        private val workflows: Map<String, Workflow>

        private val parts: List<Part>

        init {
            if (lines.isNotEmpty()) {
                val sep = lines.indexOf("")
                workflows = lines.subList(0, sep).associate { line ->
                    val (name, rules) = WORKFLOW_PATTERN.find(line)!!.destructured
                    name to Workflow(rules)
                }
                parts = lines.subList(sep + 1, lines.size).map { Part(it) }
            } else {
                workflows = mapOf()
                parts = listOf()
            }
        }

        private fun handlePart(part: Part): Boolean {
            var currentWorkflow = workflows["in"]!!
            while (true) {
                val result = currentWorkflow.process(part)
                if (result == "A")
                    return true
                if (result == "R")
                    return false
                currentWorkflow = workflows[result]!!
            }
        }

        fun sumAcceptedParts(): Int = parts.filter(::handlePart).sumOf { it.rating() }

        private fun upperBound(constraints: List<Rule>, category: Char): Int =
            constraints
                .filter { rule -> rule is LessThanRule && rule.variable == category }
                .minOfOrNull { (it as LessThanRule).value } ?: 4001

        private fun lowerBound(constraints: List<Rule>, category: Char): Int =
            constraints
                .filter { rule -> rule is GreaterThanRule && rule.variable == category }
                .maxOfOrNull { (it as GreaterThanRule).value } ?: 0

        private fun collectResult(collectedRules: List<Rule>): Long =
            "xmas".map { v -> upperBound(collectedRules, v) - lowerBound(collectedRules, v) - 1 }
                .map { it.toLong() }
                .reduce { acc, i -> acc * i }

        fun dfs(currentNode: String = "in", i: Int = 0, constraints: List<Rule> = listOf()): Long {
            if (currentNode == "A") {
                return collectResult(constraints)
            }
            if (currentNode == "R") {
                return 0
            }
            val currentWorkflow = workflows[currentNode]!!
            val currentRule = currentWorkflow.rules[i]
            // match
            var total = dfs(currentRule.nextWorkflow, 0, constraints + currentRule)
            // no match
            if (i + 1 < currentWorkflow.rules.size) {
                total += dfs(currentNode, i + 1, constraints + currentRule.anti())
            }
            return total
        }
    }

    private val workflows = Workflows(lines)

    override fun part1(): Int {
        return workflows.sumAcceptedParts()
    }

    override fun part2(): Long {
        return workflows.dfs()
    }
}
